src/main/java/com/google/sps/servlets/UserFoldersServlet.java

```diff
@@ -21,65 +24,76 @@
   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
     UserService userService = UserServiceFactory.getUserService();
 
-    if (userService.isUserLoggedIn()) {
-
-      List<Folder> userFolders = new ArrayList<>();
+    Map<String, Object> jsonInfo = new HashMap<>();
+		List<Folder> userFolders = new ArrayList<>();
+    jsonInfo.put(CREATE_FORM_HEADER, false);
 
-      String userKey = request.getParameter("userKey");
+    if (userService.isUserLoggedIn()) {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-      // Query all folders identified by the userKey
-      Query folderQuery = new Query("Folder").setAncestor(KeyFactory.stringToKey(userKey));
+      String userEmail = userService.getCurrentUser().getEmail();
+      Query userQuery = new Query("User").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, userEmail));
+			Entity userEntity = datastore.prepare(userQuery).asSingleEntity();
+      Key userKey = userEntity.getKey();
 
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Query folderQuery = new Query("Folder").setAncestor(userKey);
       PreparedQuery results = datastore.prepare(folderQuery);
 
-      // Add folders to a list
       if (results != null) {
-        for (Entity entity: results.asIterable()) {
-          Folder folder = Folder.EntityToFolder(entity);
+        for (Entity entity : results.asIterable()) {
+          Folder folder = new Folder(entity);
+          folder.setFolderKey(KeyFactory.keyToString(entity.getKey()));
           userFolders.add(folder);
         }
       }
-      
-      response.setContentType("application/json;");
-      response.getWriter().println(new Gson().toJson(userFolders));
+		
+		jsonInfo.put(CREATE_FORM_HEADER, true);
     }
+
+    jsonInfo.put(USER_FOLDER_HEADER, userFolders);
+    response.setContentType("application/json;");
+    response.getWriter().println(new Gson().toJson(jsonInfo));
   }
   
-  /**
-  ** Store new folder into Datastore
-  **/
-
+  /*
+  * Store new folder into Datastore
+  */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) {
-
     UserService userService = UserServiceFactory.getUserService();
 
     if (userService.isUserLoggedIn()) {
+			DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
       
       String folderName = request.getParameter("folderName");
       String folderDefaultLanguage = request.getParameter("folderDefaultLanguage");
-      String userKey = request.getParameter("userKey");
+
+      String userEmail = userService.getCurrentUser().getEmail();
+      Query userQuery = new Query("User").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, userEmail));
+			Entity userEntity = datastore.prepare(userQuery).asSingleEntity();
+      String userKey = KeyFactory.keyToString(userEntity.getKey());
```

Comment:
this section is the same as what starts on line 50. Rather than repeat the same code, better to separate out in a helper function

Reply:
Thanks, I just realized that there is this weird bug. When I have my file configured to two spaces for the tab button, the file has weird spacing when it is pushed. This explains why some lines are indented a lot. I'll fix this